src/libraries/System.Private.CoreLib/src/System/Text/Ascii.cs

```diff
@@ -14,56 +16,215 @@ public static partial class Ascii
+                    {
+                        if (!AllCharsInVectorAreAscii(
+                            Vector256.LoadUnsafe(ref searchSpace) |
+                            Vector256.LoadUnsafe(ref searchSpace, (nuint)Vector256<T>.Count) |
+                            Vector256.LoadUnsafe(ref searchSpace, 2 * (nuint)Vector256<T>.Count) |
+                            Vector256.LoadUnsafe(ref searchSpace, 3 * (nuint)Vector256<T>.Count)))
+                        {
+                            return false;
+                        }
+
+                        searchSpace = ref Unsafe.Add(ref searchSpace, 4 * Vector256<T>.Count);
+                    }
+
+                    searchSpace = ref finalStart;
+                }
+
+                // Process the last [1, 128] bytes.
+                // The search space has at least 2 * Vector256 bytes available to read.
+                // We process the first 2 and last 2 vectors, which may overlap.
+                return AllCharsInVectorAreAscii(
+                    Vector256.LoadUnsafe(ref searchSpace) |
+                    Vector256.LoadUnsafe(ref searchSpace, (nuint)Vector256<T>.Count) |
+                    Vector256.LoadUnsafe(ref Unsafe.Subtract(ref searchSpaceEnd, 2 * (nuint)Vector256<T>.Count)) |
+                    Vector256.LoadUnsafe(ref Unsafe.Subtract(ref searchSpaceEnd, (nuint)Vector256<T>.Count)));
+            }
+            else
+            {
+                // Process long inputs 64 bytes at a time.
+                if (length > 4 * Vector128<T>.Count)
+                {
+                    // Process the first 64 bytes.
+                    if (!AllCharsInVectorAreAscii(
+                        Vector128.LoadUnsafe(ref searchSpace) |
+                        Vector128.LoadUnsafe(ref searchSpace, (nuint)Vector128<T>.Count) |
+                        Vector128.LoadUnsafe(ref searchSpace, 2 * (nuint)Vector128<T>.Count) |
+                        Vector128.LoadUnsafe(ref searchSpace, 3 * (nuint)Vector128<T>.Count)))
+                    {
+                        return false;
+                    }
+
+                    searchSpace = ref Unsafe.Add(ref searchSpace, 4 * Vector128<T>.Count);
+
+                    // Try to opportunistically align the reads below. The input isn't pinned, so the GC
+                    // is free to move the references. We're therefore assuming that reads may still be unaligned.
+                    // They may also be unaligned if the input chars aren't 2-byte aligned.
+                    nuint misalignedElements = ((nuint)Unsafe.AsPointer(ref searchSpace) & (nuint)(Vector128<byte>.Count - 1)) / (nuint)sizeof(T);
+                    searchSpace = ref Unsafe.Subtract(ref searchSpace, misalignedElements);
+
+                    ref T finalStart = ref Unsafe.Subtract(ref searchSpaceEnd, 4 * Vector128<T>.Count);
+
+                    while (Unsafe.IsAddressLessThan(ref searchSpace, ref finalStart))
```

Comment:
You mean something like this: https://github.com/dotnet/runtime/commit/f26018aed1056e4d77f908eeb528aa7802a8ea7e?
Personally, I'm so used to the current approach that I find that harder to read.

The linked commit is `pr-4-for` (I haven't looked at codegen for any obvious tweaks that could bring some of the diff back):
|            Method |        Toolchain | Length |      Mean |     Error | Ratio | Code Size |
|------------------ |----------------- |------- |----------:|----------:|------:|----------:|
| AsciiIsValid_Byte |             pr-4 |  10000 | 54.518 ns | 0.2602 ns |  1.00 |     542 B |
| AsciiIsValid_Byte |     pr-4-noAlign |  10000 | 71.884 ns | 0.4822 ns |  1.32 |     491 B |
| AsciiIsValid_Byte |         pr-4-for |  10000 | 70.886 ns | 0.2585 ns |  1.30 |     572 B |
| AsciiIsValid_Byte | pr-4-for-noAlign |  10000 | 73.866 ns | 0.3796 ns |  1.35 |     506 B |

Reply:
Yes, although there are a few things that could be done to improve readabiliy more. We should also be able to do all this without the 16ns perf hit, it just might require slightly more.

Reply:
This is what the main loop looks like when switching to the for loop:
![image](https://user-images.githubusercontent.com/25307628/232613565-63b8cd84-43c0-4baf-96e1-ff4261e169bb.png)
It's not a large difference, but given how cheap the core checks are, doing anything extra is _relatively_ expensive.

> it just might require slightly more

How would you restructure the loop differently?

Reply:
In https://github.com/dotnet/runtime/commit/f26018aed1056e4d77f908eeb528aa7802a8ea7e I added some comments on how to make the for-loop faster.

Reply:
Thanks. I guess it just dislikes the addressing mode?

![image](https://user-images.githubusercontent.com/25307628/232624262-165ac27a-06be-46e3-ab32-5bbdad8b5325.png)

|            Method | Toolchain | Length |     Mean |    Error | Ratio | Code Size |
|------------------ |---------- |------- |---------:|---------:|------:|----------:|
| AsciiIsValid_Byte |       for |  10000 | 64.92 ns | 0.061 ns |  1.20 |     554 B |
| AsciiIsValid_Byte |        pr |  10000 | 54.12 ns | 0.184 ns |  1.00 |     542 B |

Reply:
It really seems like the complex addressing mode is worse here.
Can you try to do a Unsafe.Add at the begin of the loop body to position the `searchSpace`,  so hopefully the simple adressing mode comes back.
That positioning should result in one `lea`, so being cheap.

Reply:
> Though I remain unconvinced that [this](https://github.com/MihaZupan/runtime/commit/e8c242e75a15f6d187022cbc23b9ea9d3d83ce4e) is an improvement readability/maintainability-wise

I agree. Plus it's still 4 % slower.
The concern about the GC-hole is valid, but the trouble comes more with a backwards-loop. Here the scan is forward, so the "pointer chasing" as you had before should be safe and quite easy to review. So I'd use the approach from before here.

Note: some `nuint`-casts could be avoided by declaring the types as `nuint`, but it doesn't make it really prettier in the whole.

Reply:
> The concern about the GC-hole is valid, but the trouble comes more with a backwards-loop

You definitely run into more trouble with almost any subtraction than with addition, but the general risks exist either way. It's altogether safer to not modify the byref and we have explicitly opted to lose some small amount of perf in the past in favor of having the safer code.

Some of the readability issues come from the general setup as well and would be solved by extracting some of the key boundaries ahead of time.

For example, doing:
```csharp
nuint nativeLength = (uint)length;
(nuint blockCount, nuint remainingElements) = nuint.DivRem(nativeLength, (uint)Vector256<T>.Count);
```

Then later doing:
```csharp
for (nuint i = 0; i < blockCount; i += 4)
{
    if (!AllCharsInVectorAreAscii(
            Vector256.LoadUnsafe(ref searchSpace, i + (0 * (uint)Vector256<byte>.Count)) |
            Vector256.LoadUnsafe(ref searchSpace, i + (1 * (uint)Vector256<byte>.Count)) |
            Vector256.LoadUnsafe(ref searchSpace, i + (2 * (uint)Vector256<byte>.Count)) |
            Vector256.LoadUnsafe(ref searchSpace, i + (3 * (uint)Vector256<byte>.Count))
        ))
    {
        return false;
    }
}
```

Gives you a very nice and simple:
```asm
loop_header:
    xor r8d, r8d
    test rcx, rcx
    jbe short loop_done
    vmovupd ymm0, [0x7ffcfa5500a0]
    nop [rax+rax]

loop_iterate:
    vmovdqu ymm1, [rdx+r8]
    vpor ymm1, ymm1, [rdx+r8+0x20]
    vpor ymm1, ymm1, [rdx+r8+0x40]
    vpor ymm1, ymm1, [rdx+r8+0x60]
    vptest ymm1, ymm0
    jne short L0054
    add r8, 4
    cmp r8, rcx
    jb short loop_iterate

loop_done:
```

You can then handle trailing data starting at `nuint i = nativeLength - remainingElements;` or if idempotent from `nuint i = nativeLength - (uint)Vector256<T>.Count;`

Following such an approach also makes it much easier for non-advanced users to take their existing loops and convert them to vectorized code, as it doesn't require contorting or massively modifying the code. Instead, it follows the general principle of:
1. Write regular algorithm
2. Unroll regular algorithm to VectorXXX<T>.Count
3. Squash count operations to 1 SIMD operation

We are able to use regular for loops, reduce total casting, and write logic that is overall very similar to the original scalar algorithm.

Reply:
CC. @jkotas for input on the above.

Just another case of modifying the `byref` with `Unsafe.Add/Subtract` vs using a standard `for loop` and `LoadUnsafe(ref, nuint)`

There is a small performance hit of about 2ns for 10k elements processed (average of about 6 cycles or so, on a 3GHz CPU), which I think is more than acceptable given the additional safety and reduced risk of GC hole.

Reply:
> There is a small performance hit of about 2ns for 10k elements processed

This number was testing https://github.com/MihaZupan/runtime/commit/e8c242e75a15f6d187022cbc23b9ea9d3d83ce4e, which explicitly lifts the offset calculation out of the 4 vector loads. When using the addressing mode you mentioned, the observed difference was 20%: https://github.com/dotnet/runtime/pull/84881#discussion_r1169322845.

> ```c#
> (nuint blockCount, nuint remainingElements) = nuint.DivRem(nativeLength, (uint)Vector256<T>.Count);
> for (nuint i = 0; i < blockCount; i += 4)
> ```

A few notes:
- How would you incorporate the alignment step with this approach?
- You still need a different loop condition, e.g. `i <= blockCount - 4`
- This PR handles the remaining 2-4 vectors worth of data with [the same logic](https://github.com/dotnet/runtime/blob/9a97e6386da6b35ff5e3a3808f23ea7e877d700e/src/libraries/System.Private.CoreLib/src/System/Text/Ascii.cs#L171-L175). That is, the same logic is used at the end if you had 3 vectors worth of data or 5 or 10. With your `blockCount` approach, you now need to either give that up, or have some other way of communicating where you can start reading the input from. 
- The offset calculation in your example is wrong, you probably meant `LoadUnsafe(ref searchSpace, (i + 0) * (uint)Vector256<byte>.Count)`, but that ends up with
  ```asm
  vmovups   ymm0,[rcx+r10]
  lea       r10,[r9+1]
  shl       r10,5
  vpor      ymm0,ymm0,[rcx+r10]
  lea       r10,[r9+2]
  shl       r10,5
  vpor      ymm0,ymm0,[rcx+r10]
  lea       r10,[r9+3]
  shl       r10,5
  vpor      ymm0,ymm0,[rcx+r10]
  vptest    ymm0,ymm1
  ```
  so you're back at incrementing by `4 * Vector256<T>.Count` or using some other approach

I'm okay with changing the implementation if you believe it will promote using less error-prone approaches in general, but I'm pointing out that it's not as simple as just using a for loop (it ends up being closer to [this](https://github.com/MihaZupan/runtime/commit/e8c242e75a15f6d187022cbc23b9ea9d3d83ce4e), which I dislike), and that the performance differences involved are larger than 2 ns.

Reply:
> When using the addressing mode you mentioned, the observed difference was 20%

Ah, right. Misread. Still, 10ns for 10k elements, which is on average around 30 additional cycles seems reasonable overall.

> The offset calculation in your example is wrong, you probably meant LoadUnsafe(ref searchSpace, (i + 0) * (uint)Vector256<byte>.Count), but that ends up with

Right, sorry, mixed up two different approaches in my head vs when writing it down and forgot to factor in the vector unrolling:
```csharp
nuint nativeLength = (uint)length;
nuint remainingElements = nativeLength % (4 * (uint)Vector256<T>.Count);
nuint blockEnd = nativeLength - remainingElements;    
```

and
```csharp
        for (nuint i = 0; i < blockEnd; i += (4 * (uint)Vector256<T>.Count))
```

> How would you incorporate the alignment step with this approach?

If you're opportunistically trying to align data (without pinning) and the operation is idempotent, then you can do something like:
```csharp
nuint nativeLength = (uint)length;

nuint address = (nuint)Unsafe.AsPointer(ref searchSpace);
nuint misalignment = address % (uint)Vector256<byte>.Count;

nuint blockStart = 0;

// Checking that the misalignment is safe is important
if ((misalignment & (uint)(sizeof(T) - 1)) == 0)
{
    // TODO: Handle one misaligned vector (not one iteration)
    blockStart += misalignment;
}

nuint remainingElements = (nativeLength - blockStart) % (4 * (uint)Vector256<T>.Count);
nuint blockEnd = nativeLength - remainingElements;

for (nuint i = blockStart; i < blockEnd; i += (4 * (uint)Vector256<T>.Count))
{
}

if (remainingElements != 0)
{
    // TODO: Handle one iteration starting at `nativeLength - 4 * Vector256<T>.Count`
}
```

There are, of course, other ways to do this as well, but the premise remains the same and you can still easily handle alignment, trailing elements, etc.

> You still need a different loop condition, e.g. i <= blockCount - 4

This all depends on how you structure your logic. `blockEnd` is already going to be definitely in bounds and so you have the option of just checking `remainingElements is not zero` (which is a fused `tst, jz`). You could alternatively say that you always want to process trailing elements without branching and account for that in `blockEnd` up front, but that can also end up with weird scenarios and the single fused branch is often just as cheap as the additional subtraction logic

Reply:
> CC. @jkotas for input on the above.

I agree that it would be nice to prefer less bug prone patterns. Non-mutable byref + indexing is less bug prone than mutable byrefs. In my opinion, pinning the buffer and using regular unmanaged pointers would be just fine for more complex leaf methods like this one (short term pinning is not a problem for GC).

We are chasing nanoseconds in microbenchmarks and argue about this on one case-by-case basis, so CoreLib does not have consistent pattern for this sort of code.